starknetdata/gateway/gateway.go

```diff
@@ -0,0 +1,223 @@
+package gateway
+
+import (
+	"errors"
+
+	"github.com/NethermindEth/juno/clients"
+	"github.com/NethermindEth/juno/core"
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/juno/utils"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+type Gateway struct {
+	client *clients.GatewayClient
+}
+
+func NewGateway(n utils.Network) *Gateway {
+	return &Gateway{
+		client: clients.NewGatewayClient(n.URL()),
+	}
+}
+
+// BlockByNumber gets the block for a given block number from the feeder gateway,
+// then adapts it to the core.Block type.
+func (g *Gateway) BlockByNumber(blockNumber uint64) (*core.Block, error) {
+	response, err := g.client.GetBlock(blockNumber)
+	if err != nil {
+		return nil, err
+	}
+
+	return AdaptBlock(response)
+}
+
+func AdaptBlock(response *clients.Block) (*core.Block, error) {
+	if response == nil {
+		return nil, nil
+	}
+
+	// Receipts
+	receipts := make([]*core.TransactionReceipt, len(response.Receipts))
+	for i, receipt := range response.Receipts {
+		receipts[i] = adaptTransactionReceipt(receipt)
+	}
+
+	// Events
+	eventCommitment, eventCount, err := core.EventCommitmentAndCount(receipts)
+	if err != nil {
+		return nil, err
+	}
+
+	// Transaction Commitment
+	txCommitment, err := core.TransactionCommitment(receipts)
+	if err != nil {
+		return nil, err
+	}
+
+	return &core.Block{
+		Hash:                  response.Hash,
+		ParentHash:            response.ParentHash,
+		Number:                response.Number,
+		GlobalStateRoot:       response.StateRoot,
+		Timestamp:             new(felt.Felt).SetUint64(response.Timestamp),
+		TransactionCount:      new(felt.Felt).SetUint64(uint64(len(response.Transactions))),
+		TransactionCommitment: txCommitment,
+		EventCount:            new(felt.Felt).SetUint64(eventCount),
+		EventCommitment:       eventCommitment,
+		ProtocolVersion:       new(felt.Felt),
+		ExtraData:             nil,
+	}, nil
+}
+
+func adaptTransactionReceipt(response *clients.TransactionReceipt) *core.TransactionReceipt {
+	if response == nil {
+		return nil
+	}
+
+	events := make([]*core.Event, len(response.Events))
+	for i, event := range response.Events {
+		events[i] = adaptEvent(event)
+	}
+
+	l2ToL1Messages := make([]*core.L2ToL1Message, len(response.L2ToL1Message))
+	for i, msg := range response.L2ToL1Message {
+		l2ToL1Messages[i] = adaptL2ToL1Message(msg)
+	}
+
+	return &core.TransactionReceipt{
+		ActualFee:          response.ActualFee,
+		TransactionHash:    response.TransactionHash,
+		TransactionIndex:   response.TransactionIndex,
+		Events:             events,
+		ExecutionResources: adaptExecutionResources(response.ExecutionResources),
+		L1ToL2Message:      adaptL1ToL2Message(response.L1ToL2Message),
+		L2ToL1Message:      l2ToL1Messages,
```

Comment:
Good spot! I think this needs to be discussed in conjunction with the above comment and it is out of the scope of this pr. If it were to be fixed here it would require too many changes.

Reply:
I think we should implement the adapter in a separate PR. That way, we can also update the Block to include transaction and transaction receipt. 

We also won't need to add the transaction type to transaction receipts